reactioncommerce/reaction
# Configurable stripe application fee

### What this PR accomplishes
This PR resolves #2771 by doing the following:
* adding a new input in the Stripe Payment settings where a marketplace manager can input a `Percentage Application Fee` which is used in place of the previously hard coded 20% demo application fee.
  <img src="https://user-images.githubusercontent.com/1203639/30360569-7f3c2f70-980f-11e7-8d19-06e2982120ea.png" width="50%"/>
* Adding a new hook `onCalculateStripeApplicationFee` that fires whenever a stripe application fee is calculated.
  - This hook is able to adjust the application fee charged on a shop-by-shop basis.
  - This hook will fire once for each charge. If a checkout has 3 merchant shops represented, the hook will fire three times, once for each merchant.
 - This hook has access to the full `cart` which is being processed for the order, and the `shopId` for which currently creating a connect charge.

### Documentation
Here's the jsdoc for the hook which will 
```
/**
 * Hook for affecting the application fee charged. Any hooks that `add` "onCalculateStripeApplicationFee" will
 * run here
 * @module hooks/payments/stripe
 * @method onCalculateStripeApplicationFee
 * @param {number} applicationFee the exact application fee in cents (must be returned by every hook)
 * @param {object} options object containing properties passed into the hook
 * @param {object} options.cart the cart object
 * @param {object} options.shopId the shopId
 * @todo Consider abstracting the application fee out of the Stripe implementation, into core payments
 * @returns {number} the application fee after having been through all hooks (must be returned by ever hook)
 */
```

### Example
From a server file in your custom plugin

As an example of how this works I created a custom plugin in my `imports/plugins/custom` directory.
Here's what the structure of the plugin looks like:
![image](https://user-images.githubusercontent.com/1203639/30445176-750c4b46-9942-11e7-906a-81cd8b2a64b3.png)

The meat of the code is in the `app-fee.js` file.

#### `my-plugin/server/hooks/app-fee.js`
```
import { Hooks } from "/server/api";

function calculateStripeConnectApplicationFee(applicationFee, options) {
  const { cart, shopId } = options;
  applicationFee = 10; // Adjust the applicationFee to be a flat charge of 10 cents
  return applicationFee;
}

// Run calculateStripeConnectApplicationFee for the onCalculateStripeApplicationFee hook
Hooks.Events.add("onCalculateStripeApplicationFee", calculateStripeConnectApplicationFee);
```

For completeness here's the full content of the other files in this demo plugin.
#### `my-plugin/server/hooks/index.js`
```
import "./app-fee";
```

#### `my-plugin/server/index.js`
```
import "./hooks";
```

#### `my-plugin/register.js`
```
import { Reaction } from "/server/api";

Reaction.registerPackage({
  label: "My Plugin",
  name: "namespace-my-plugin",
  icon: "fa fa-grav",
  autoEnable: true,
  settings: {
  },
  registry: []
});

```


### To Test
1. Setup a reaction marketplace with at least one merchant shop.
  - setup stripe connect on the primary shop
    - Set the application fee in the Stripe settings to some number (number represents percentage fee)
  - setup at least one shipping provider on the primary shop
  - invite a merchant shop
  - accept invite, and setup stripe connect as merchant shop
  - create and publish a product as merchant shop.

2. Add 1 Basic Reaction Product and 1 of your merchant shop's products to your cart
  - You only need the primary shop's product until #2836 is resolved
3. Checkout
  - Make sure you create a user with an email address to checkout, as stripe connect requires an email
4. Visit your Merchant Stripe Account, you can do this by clicking the Stripe Account Link from `My Shop Settings` section of Shops dashboard
  ![image](https://user-images.githubusercontent.com/1203639/30445704-1f771646-9944-11e7-9005-fd01f6218988.png)
5. Navigate to the payments section of your stripe dashboard
![image](https://user-images.githubusercontent.com/1203639/30445856-a60adcb0-9944-11e7-82b0-69e4cc92091d.png)
6. Click on your payment that just came through
7. Click Capture on that payment
  ![image](https://user-images.githubusercontent.com/1203639/30445902-cb95a7e4-9944-11e7-8519-560b85bf9eab.png)
8. Note the application fee by hovering over the Fee helper icon
  ![image](https://user-images.githubusercontent.com/1203639/30446005-2033a4ea-9945-11e7-8934-28be93280785.png)

#### To test the hook
- Do all of the above steps first
- Create a new plugin similar to the plugin outlined in this PR, and then run through another order.
- When you capture the charge, you'll see that it's application fee is from the plugin and not the dashboard input field.
  ![image](https://user-images.githubusercontent.com/1203639/30446108-763376c2-9945-11e7-8e91-861338f5cac6.png)
